src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -798,3 +832,63 @@ mergeRuns resolve hfs hbio caching alloc runPaths mergeLevel runs = do
     Merge.new hfs hbio caching alloc mergeLevel resolve runPaths runs >>= \case
       Nothing -> error "mergeRuns: no inputs"
       Just m -> Merge.stepsToCompletion m 1024
+
+{-------------------------------------------------------------------------------
+  Credits
+-------------------------------------------------------------------------------}
+
+type Credit = Int
+
+{-# SPECIALISE supplyCredits ::
+     Credit
+  -> Levels IO h
+  -> IO ()
+  #-}
+supplyCredits ::
+     (MonadSTM m, MonadST m, MonadMVar m, MonadMask m, MonadFix m)
+  => Credit
+  -> Levels m h
+  -> m ()
+supplyCredits c levels =
+    V.iforM_ levels $ \_i (Level mr _rs) ->
+    -- let !ln = i + 1 in
+    let !cr = creditsForMerge mr in
+    supplyMergeCredits (ceiling (fromIntegral c * cr)) mr
+
+creditsForMerge :: MergingRun m h -> Rational
+creditsForMerge SingleRun{}                             = 0
+creditsForMerge (MergingRun LevelLevelling _ _)         = 1 + 4
+creditsForMerge (MergingRun LevelTiering (NumRuns n) _) = fromIntegral n / 4
+
+{-# SPECIALISE supplyMergeCredits ::
+     Credit
+  -> MergingRun IO h
+  -> IO ()
+  #-}
+-- TODO: implement doing merge werk in batches, instead of always taking the
+-- MVar. The thresholds for doing merge work should be different for each level,
+-- maybe co-prime?
```

Comment:
We should implement this in a follow-up PR. The current version without batching shows that our naive implementation (without batching) already works, and then we can optimise by including batching

Reply:
See #435